server/game/effects.js

```diff
@@ -1452,7 +1452,7 @@ const Effects = {
             targetType: 'player',
             apply: function(player, context) {
                 const shadows = player.shadows;
-                const revealFunc = reveal => shadows.includes(reveal);
+                const revealFunc = reveal => player.shadows.includes(reveal);
```

Comment:
Something gone wrong. I reverted all my changes and switched to using `let` instead of `const` in the declaration of "shadows":

`let shadows = player.shadows;`

but then the bug riappeared.

I think a possible reason is: `const` cannot be reassigned once assigned, but the referenced value can be changed (if not immutable, like strings). For example:

`var data = [0, 1, 2];`
`const constData = data; //constData is [0, 1, 2]`
`data.push(3); //now constData is [0, 1, 2, 3] as well`
`constData.push(4); //now both data and constData are [0, 1, 2, 3, 4]`
`constData = []; // trying to reassign a new value raise an exception "Uncaught TypeError: Assignment to constant variable."`

Maybe here (but it is only a supposition) is happening a reassignment of `player.shadows` attribute. Something like this:

`const shadows = player.shadows; //both shadows and player.shadows refers to the same array`
`const revealFunc = reveal => shadows.includes(reveal); // revealFunc is bound to the array referenced by const shadows`
`// ...`
`// now my hypothesis is that, in some other part of the code, the player.shadows attribute is reassigned. Something like this:`
`player.shadows = player.shadows.filter(() => {....}); `
`// the above instruction creates a new array (shallow copy) and now the array that was previously referenced by player.shadows and by const shadows is different from the array that is referenced by player.shadows. The result is that changes made to player.shadows are not reflected to the const shadows variable... but the revealFunc is bound to that variable. Continuing my previous example, maybe is happening something like this:`

`//both data and constData are [0, 1, 2, 3, 4]`
`var dataB = data.filter(() => {return true;}); //dataB is [0, 1, 2, 3, 4] but is stored in a different "memory area"`
`dataB.push(5); //now dataB is [0, 1, 2, 3, 4, 5] but data and constData are [0, 1, 2, 3, 4]`
`data.push(6); //dataB is [0, 1, 2, 3, 4, 5] and data and constData are [0, 1, 2, 3, 4, 6]. Values have diverged.`

If my hypothesis is correct, this is also the reason why my previous fix actually works. 
By bounding the `revealFunc` to the `player` object, we are able to access the `player.shadows` attribute when the `revealFunc` is executed, reading from the actual "memory area" referenced by `player.shadows`.

Last note: the `const shadows = player.shadows;` declaration remains unchanged, so line 1460 (and others) are safe. The only change is in the definition of the `revealFunc` function, so there should be no other side effect in my opinion.

Let me know if this make sense for you :)

About test case: absolutely yess. I noticed other test cases, and I would love to do test cases for this case (and maybe also for the other two fixes, if you are ok with it). I will try to write test case by myself, eventually I will contact you :)

Reply:
Good investigation. I think you're right in your suspicion of `player.shadows` being reassigned; it seems that the method which handles cards being removed from piles does exactly this within [player.js](https://github.com/Patane97/throneteki/blob/15c772813e6ae0718239a9f47c58847ed7949231/server/game/player.js#L840-L842), which is called when a card is removed from a pile (which I believe is an atomic action of cards leaving shadows, but haven't 100% confirmed this).

That being the case, you're correct in your original adjustment, however the `const shadows` needs to be accounted for; we require this to track which cards are _currently_ being revealed by this effect [here](https://github.com/Patane97/throneteki/blob/15c772813e6ae0718239a9f47c58847ed7949231/server/game/effects.js#L1458-L1461) so that, within the "reapply" of that effect (which gets called whenever the game-state changes), it only applies the "revealCards" GameAction to _new_ cards which have entered shadows; therefore it will need the current list of shadow cards from the last time it was applied or reapplied. Note that the "revealCards" GameAction is ultimately what's responsible for triggering interrupts & reactions to cards being revealed, such as Sweetrobin.

Since `player.shadows` is always being reassigned, we likely should just remove `const shadows = player.shadows`, and always check `player.shadows` directly in the `apply` and `reapply` functions of this effect.

Happy to explain further if you need, as this relatively new reveal code can get a bit strange at times 😄.

Reply:
I haven't had the chance to physically test it, but I'm still not sure that the change to what's stored in `context.revealShadows[player.name]` is correct; that **should** be essentially capturing a "snapshot" of that players shadow cards at the moment of **apply**ing or **reapply**ing the effect (essentially doing what the initial issue for the `revealFunc` was). When the effect is reapplied, it uses `context.revealShadows` to know what cards have already been revealed via GameActions.revealCards in a previous apply/reapply, and thus should not be revealed again. If we're storing the player reference in `context.revealShadows[player.name]`, then `player.shadows` will always be equal to `context.revealShadows[player.name].controller.shadows` when the effect is being reapplied, so it won't detect any changes & won't trigger `GameActions.revealCards` on any new cards, I think.

I can give this an actual test sometime tomorrow at work maybe to confirm if I'm right there. :)

The test looks great though!